pkg/rest/rate_limit.go

```diff
@@ -0,0 +1,131 @@
+/**
+ * @license
+ * Copyright 2020 Dynatrace LLC
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package rest
+
+import (
+	"errors"
+	"github.com/dynatrace-oss/dynatrace-monitoring-as-code/pkg/util"
+	"net/http"
+	"time"
+)
+
+// rateLimitStrategy ensures that the concrete implementation of the rate limiting strategy can be hidden
+// behind this interface
+type rateLimitStrategy interface {
+	executeRequest(timelineProvider util.TimelineProvider, callback func() Response) (Response, error)
+}
+
+// createRateLimitStrategy creates a rateLimitStrategy. In the future this can be extended to instantiate
+// different rate limiting strategies based on e.g. environment variables. The current implementation
+// always returns the strategy simpleSleepRateLimitStrategy, which suspends the current goroutine until
+// the time in the rate limiting header 'X-RateLimit-Reset' is up.
+func createRateLimitStrategy() rateLimitStrategy {
+	return &simpleSleepRateLimitStrategy{}
+}
+
+// simpleSleepRateLimitStrategy, is a rate limiting strategy which suspends the current goroutine until
+// the time in the rate limiting header 'X-RateLimit-Reset' is up.
+// It has a min sleep duration of 5 seconds and a max sleep duration of one minute and performs maximal 5
+// polling iterations before giving up.
+type simpleSleepRateLimitStrategy struct{}
+
+func (s *simpleSleepRateLimitStrategy) executeRequest(timelineProvider util.TimelineProvider, callback func() Response) (Response, error) {
+
+	response := callback()
+
+	maxIterationCount := 5
+	currentIteration := 0
+
+	for response.StatusCode == http.StatusTooManyRequests && currentIteration <= maxIterationCount {
+
+		limit, humanReadableTimestamp, timeInMicroseconds, err := s.extractRateLimitHeaders(response)
```

Comment:
No, since Dynatrace returns all timestamps in UTC. I'm guessing the rate limiting headers are no exception here. But  now that I think of it... you are right - the `timelineProvider` needs to return UTC as well, I'll check the code again. Great input!

Reply:
Done in 9cda9ba